MGDSTRM-8899: update README and refactor (#41)
bf2fc6cc711aee1a0c2a/kas-broker-plugins: README.adoc

- `io.bf2.kafka.authorizer.CustomAclAuthorizer` - an https://kafka.apache.org/documentation/#security_authz[Authorizer] implementation
- `io.bf2.kafka.topic.ManagedKafkaCreateTopicPolicy` - a https://cwiki.apache.org/confluence/display/KAFKA/KIP-108%3A+Create+Topic+Policy[CreateTopicPolicy] implementation

> - `io.bf2.kafka.config.ManagedKafkaAlterConfigPolicy` - a https://cwiki.apache.org/confluence/display/KAFKA/KIP-133%3A+Describe+and+Alter+Configs+Admin+APIs#KIP133:DescribeandAlterConfigsAdminAPIs-AlterConfigs[AlterConfigPolicy] implementation

Reviewer: ```suggestion
- `io.bf2.kafka.config.ManagedKafkaAlterConfigPolicy` - an https://cwiki.apache.org/confluence/display/KAFKA/KIP-133%3A+Describe+and+Alter+Configs+Admin+APIs#KIP133:DescribeandAlterConfigsAdminAPIs-AlterConfigs[AlterConfigPolicy] implementation
```

When no rules match, then by default it is denied.
So, if ACL is not defined for a resource type then it automatically gets denied.

== ManagedKafkaCreateTopicPolicy

The objective of this policy is to:

* validate the topic replication factor (default allows 3 only)
* validate number of partitions in the cluster
* validate the create topic configs

> To config the create topic policy, you should add config in Strimzi's Kafka CR.

Reviewer 2: I think I would pull this out as its own section. `Configuring the policies`

      alter.config.policy.class.name=io.bf2.kafka.config.ManagedKafkaAlterConfigPolicy
      # partition limit setting
      kas.policy.create-topic.max.partitions=1000
      kas.policy.create-topic.partition-counter.timeout-seconds=10
      kas.policy.create-topic.partition-counter.private-topic-prefix="__redhat"
      kas.policy.create-topic.partition-counter.schedule-interval-seconds=15
      kas.policy.create-topic.partition-limit-enforced=true
      # topic config setting
      kas.policy.topic-config.enforced=compression.type:producer,segment.jitter.ms:0
      kas.policy.topic-config.mutable=cleanup.policy,delete.retention.ms,retention.bytes,retention.ms
      kas.policy.topic-config.range=segment.bytes:52428800:,segment.ms:600000:
----

Here are the configuration description:

* `kas.policy.create-topic.max.partitions`: Used to specify the upper limit of partitions that should be allowed in the cluster. If this property is not specified, a default of -1 (no limit) is set
* `kas.policy.create-topic.partition-counter.timeout-seconds`: Used to specify the number of seconds to use as a timeout duration when listing and describing topics method. Default is 10.

> * `kas.policy.create-topic.partition-counter.private-topic-prefix`: Used to specify the topic prefix to match for private/internal topics. Default is no prefix.

Reviewer 2: In general `Used to specify` -> `Specifies`.

```suggestion
* `kas.policy.create-topic.partition-counter.private-topic-prefix`: Specifies the prefix  used to match against topic name to detect private/internal topics. Defaults to blank - no prefix.
```

> * `kas.policy.create-topic.partition-counter.schedule-interval-seconds`: Used to specify the interval (in seconds) at which to schedule partition counts. Default is 15.

Reviewer 2: ```suggestion
* `kas.policy.create-topic.partition-counter.schedule-interval-seconds`: Specifies the interval (in seconds) between partition counts. Defaults to 15.
```

> * `kas.policy.create-topic.partition-limit-enforced`: Used to allow disabling of partition limit enforcement. Default is false.

Reviewer 2: ```suggestion
* `kas.policy.create-topic.partition-limit-enforced`: `true` disables the enforcement of the partition limit. Defaults to `false` (limit enforced).
```

> * `kas.policy.topic-config.enforced`: Used to specify the configs with permitted value. It's a list in the form [keyA]:[valueA],[keyB]:[valueB]. Default value is: "flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"

Reviewer 2: ```suggestion
* `kas.policy.topic-config.enforced`: Specifies a mapping between a topic policy key and a value. The property is expressed as a comma separated list of `key:value` pairs where the key and value are separated by  a `:`. Hence the it takes the form `keyA:value1,keyB:value2`. Default value is: "flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"
```

I found the use of `[]` around the key and value in the format description confusing as its trying to describe the parsing rules including extra tokens felt misleading. I'm easily persuaded that everyone knows they imply optional and thus should be included...

Author: Good suggestion.

> * `kas.policy.topic-config.mutable`: Used to specify the configs that allow to be updated. It's a comma separated list. We'll merge everything in kas.policy.topic-config.enforced and kas.policy.topic-config.range config. Default value is: "message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"

Reviewer 2: ```suggestion
* `kas.policy.topic-config.mutable`: Specifies which properties of topic are allowed to be configured. It's a comma separated list. The list will be merged with `kas.policy.topic-config.enforced` and `kas.policy.topic-config.range config`. Defaults to: "message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
```

How does the merge work? IIRC an entry appearing in `kas.policy.topic-config.enforced` is enough but the idea of merging could be read as implying that you need to list all config keys here (the super set of config keys) and then the enforcement detail comes from the additional keys.

> * `kas.policy.topic-config.range`: Used to specify the configs that allow values within a range with the format "configA:minA:maxA,configB:minB:maxB,...". Either min or max can be empty, which means we only set lower/upper bound. Default value is: "max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:"

Reviewer 2: ```suggestion
* `kas.policy.topic-config.range`: Specifies a mapping from config key to an inclusive/closed range off values. Entries are formatted as a triplet using `:` as a separator, `key:min:max` where either min or max are optional implying a single bound as appropriate. Multiple triplets are comma separated. For example  `configA:minA:maxA,configB:minB:maxB,...`. Defaults to: "max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:"
```

== ManagedKafkaAlterConfigPolicy

The objective of this policy is to validate the alter topic configs

> The configuration and the description is the same as ManagedKafkaCreateTopicPolicy, except ManagedKafkaAlterConfigPolicy only accepts topic-config configs, i.e.

Reviewer 2: ```suggestion
ManagedKafkaAlterConfigPolicy shares its configuration with the ManagedKafkaCreateTopicPolicy, that is to say there is one shared policy definition. The alter topic policy only enforces the configuration defined in the following
```

I'm not 100% convinced about this form of words.

When no rules match, then by default it is denied.
So, if ACL is not defined for a resource type then it automatically gets denied.

== ManagedKafkaCreateTopicPolicy

The objective of this policy is to:

* validate the topic replication factor (default allows 3 only)
* validate number of partitions in the cluster
* validate the create topic configs

Here are the configuration description:

[cols="2,1,5"]
|===

> |property |default value |description

Reviewer 2: Personally I'd have 
```suggestion
|property |description |default value
```

Reviewer 2: ```suggestion
|property |description |default value  | default interpretation 
```
re Reading the descriptions. They all describe what the default is, duplicating the column, and generally what that means. for example from `kas.policy.create-topic.max.partitions`: `If this property is not specified, the default of ``-1`` implies no limit.`. It might be clearer to express that using 4 columns. 

```
|kas.policy.create-topic.max.partitions | Specifies the upper limit of partitions that should be allowed in the cluster.| -1 | If this property is not specified, the default of `-1` implies no limit.
```

|kas.policy.create-topic.max.partitions |-1 |Specifies the upper limit of partitions that should be allowed in the cluster. If this property is not specified, the default of `-1` implies no limit.
|kas.policy.create-topic.partition-counter.timeout-seconds |10 |Specifies the duration in seconds to use as a timeout when listing and describing topics. Defaults to 10.
|kas.policy.create-topic.partition-counter.private-topic-prefix |"" |Specifies the prefix used to match against topic name to detect private/internal topics. Defaults to blank - no prefix.
|kas.policy.create-topic.partition-counter.schedule-interval-seconds |15 |Specifies the interval (in seconds) between partition counts. Defaults to 15.

> |kas.policy.create-topic.partition-limit-enforced |false |`true` disables the enforcement of the partition limit. Defaults to `false` (limit enforced).

Reviewer 2: Is this right, the description is the opposite mapping to what I expect from the property name. 

I expect `true` -> `Enforced`, `false` -> `Disabled` but your description is the opposite.

Author: You're right! Updated.

|kas.policy.topic-config.mutable |"message.timestamp.difference.max.ms,message.timestamp.type,..."[2] |Specifies which properties of topic are allowed to be configured. It's a comma separated list. The list will be merged with `kas.policy.topic-config.enforced` and `kas.policy.topic-config.range config`.
|kas.policy.topic-config.range |"max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:" |Specifies a mapping from config key to an inclusive/closed range off values. Entries are formatted as a triplet using `:` as a separator, `key:min:max` where either min or max are optional implying a single bound as appropriate. Multiple triplets are comma separated. For example  `configA:minA:maxA,configB:minB:maxB,...`.
|===

[1]"flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"

> [2]"message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"

Author: Put these 2 value out of the table, to avoid occupying too many table width size.

Reviewer 2: Re-ordering the columns so the description is before the default might make that less of an issue.

=== Configuring the policies

> To config the create topic policy, you should add config in Strimzi's Kafka CR.

Reviewer 2: ```suggestion
To configure the create topic policy, you should add config in the Kafka CR.
```
